Bau - Update readme and add overview documentation (#164)
alphagov/verify-proxy-node: doc/overview.md

# GOV.UK Verify UK Proxy Node

## Technical overview

> The UK Proxy Node will allow GOV.UK Verify to provide identities as part of the eIDAS scheme. This document describes the architecture and behaviour of the UK Proxy Node.

Reviewer: ```suggestion
The UK Proxy Node allows GOV.UK Verify to provide identities as part of the eIDAS scheme. This document describes the architecture and behaviour of the UK Proxy Node.
```
Can just use the present tense here as well

### Role of the UK Proxy Node
As part of the eIDAS scheme, the UK must host a proxy node to allow users access to GOV.UK Verify when in an EU country. As defined in the eIDAS interoperability architecture, the proxy node will act as a front to one or more of a country’s identity schemes. 

The UK Proxy Node will be introduced into the architecture, as a REST-based microservice, to conform with eIDAS specifications. 

> GOV.UK Verify uses a SAML message format that is different from the eIDAS specification. The UK Proxy Node will support the eIDAS SAML Profile and will act as an identity provider to EU member state services and as a Relying Party to GOV.UK Verify. 

Reviewer: ```suggestion
GOV.UK Verify uses a SAML message format that is different from the eIDAS specification. The UK Proxy Node supports the eIDAS SAML Profile and acts as an identity provider to EU member state services and as a Relying Party to GOV.UK Verify. 
```

The role of the UK Proxy Node is to:
 - convert eIDAS SAML to Verify SAML messages and vice versa
 - validate all the incoming requests and outgoing responses 
 - cryptographically sign all outgoing responses to EU services

> 

Reviewer 2: Can we add to Proxy Node role
 - encrypt the identity assertions sent to the EU service

Author: Sure. I'll add that.

For requests from a connector node, the Gateway service forwards the eIDAS SAML AuthnRequest to the Eidas Saml Parser (ESP) service to perform the SAML validation. The Gateway will then use the Verify Service Provider (VSP) to generate SAML AuthnRequests which are passed back to the Gateway service before the SAML AuthnRequest is sent to the Hub.  

For responses coming from the GOV.UK Verify Hub, the Gateway service forwards them to the Translator service. Responses from the Translator service are passed on to the EU member state connector node.

### Eidas-Saml-Parser

The Gateway Service sends the SAML AuthnRequest to the Eidas Saml Parser (ESP) service to validate the data and the signature of the request. 

The ESP is the only service which can consume the countries connector metadata. The ESP will extract the Public Encryption Certificate and Destination of the Country from the metadata and send it back to the Gateway along with the RequestID and Issuer which is extracted from the SAML AuthnRequest.

### Translator

Part of the UK Proxy Node, the Translator service receives the GOV.UK Verify Hub responses forwarded by the Gateway service. 

> The Translator sends the SAML AuthnResponse to the Verify Service Provider (VSP) to perform the SAML validation. The Translator service will generate an eIDAS AuthnResponse based on the attributes it receives from the VSP. The Translator service then encrypts the signed assertion for the connector node and sends the whole eIDAS SAML response for signing by the Signing service.

Reviewer 3: What's the "Signing Service"?

Author: Was referring to the CloudHSM client. As discussed, going to remove to prevent confusion.

Once the eIDAS SAML response is signed, the Translator service passes it on to the Gateway service.

### Verify-Service-Provider

The Verify Service Provider (VSP) generates and translates Security Assertion Markup Language (SAML) messages to and from the GOV.UK Verify Hub. 

> The Gateway service uses the VSP to generate SAML requests to send to the Verify Hub. For responses, the Translator service uses the VSP to validate SAML responses and translate it to JSON. 

Reviewer: ```suggestion
The Gateway service uses the VSP to generate SAML requests to send to the Verify Hub. For responses, the Translator service uses the VSP to validate SAML responses and translate them to JSON. 
```

### Signing-Service

This component is a microservice running on an Amazon EC2 instance. The service will use an Amazon hardware security module (HSM) to sign outbound eIDAS SAML messages. 

## Connecting to Countries

> There be one instance of the proxy node services and infrastructure isolation per requesting Country connector node. They will share a single hardware security module (HSM) but will each have independent credentials to access the HSM and an independent signing key held by the HSM.

Reviewer: ```suggestion
We will host one instance of the proxy node services and infrastructure isolation per requesting Country connector node. All instances will share a single hardware security module (HSM) but will each have independent credentials to access the HSM, and an independent signing key held by the HSM.
```

## Infrastructure

The eIDAS Proxy Node is hosted on an Amazon AWS cloud infrastructure and runs on Kubernetes. This is isolated from all Verify Hub deployments to ensure that it is treated as a separate component. Public and private subnets are used to isolate the Translator and Signing services. Access to environments and deployments is controlled using AWS’s identity and access management permissions.

## Security operations monitoring

Security operations within the organisation will monitor transactions and access. Transaction monitoring will involve correlating messages entering with messages exiting the proxy node. Access monitoring will be put in place using Amazon's CloudWatch logging. Logs will be monitored by a Security Operations team to track access to servers, deployments and key rotations.

## UK Proxy Node key management

SAML responses going to EU member state connector nodes will be signed with keys stored on an HSM. These are keys with self-signed certificates. The same key will be used to sign the proxy node metadata. 

> The Verify Service Provider will be responsible for Signing AuthnRequests and Encrypting Responses from the Verify Hub. These will be stored as sealed secrets. 

Author: Yep. Will change.

| 			Key type           | 		Key location        |
| ---------------------------- | -------------------------- |

> | Metadata signing key         | AWS Key Management System  |

Author: Yep it should be. Will change.

> | AuthnRespone signing key     | AWS Key Management System  |

Author: Good spot

> | AuthnRequest signing key     | Verify Service Provider    |

Author: Yes. That would be clearer.

> | AuthnRequest signing key     | Verify Service Provider    |

Reviewer 2: I'm not sure from this how the EU member state will validate the signature created by the CloudHSM signing keys.
Do we publish the CloudHSM signing certs. I might have missed something from above ☝️ though

Author: We publish the signing cert in the Proxy Node Metadata, which they will consume.
